vpr/src/place/cut_spreader.cpp

```diff
@@ -0,0 +1,857 @@
+                    std::vector<std::pair<t_pl_loc, ClusterBlockId>> swaps_made;
+                    std::queue<std::pair<ClusterBlockId, t_pl_loc>> visit;
+                    visit.emplace(blk, sub_t);
+                    while (!visit.empty()) {
+                        ClusterBlockId vb = visit.front().first;
+                        VTR_ASSERT(!isPlaced(vb));
+                        t_pl_loc target = visit.front().second;
+                        visit.pop();
+                        auto log_type = clb_nlist.block_type(blk);
+                        auto result = std::find(log_type->equivalent_tiles.begin(), log_type->equivalent_tiles.end(), g_vpr_ctx.device().grid[target.x][target.y].type);
+                        if (result == log_type->equivalent_tiles.end()) goto fail;
+                        ClusterBlockId bound = place_ctx.grid_blocks[target.x][target.y].blocks[target.sub_tile];
+                        if (bound != EMPTY_BLOCK_ID)
+                            if (imacro(bound) != -1) goto fail;
+                        targets.emplace_back(vb, target);
+                        if (macro_head(vb) == vb) {
+                            const std::vector<t_pl_macro_member>& members = place_ctx.pl_macros[imacro(blk)].members;
+                            for (auto member = members.begin() + 1; member != members.end(); ++member) {
+                                t_pl_loc mloc = target + member->offset;
+                                visit.emplace(member->blk_index, mloc);
+                            }
+                        }
+                    }
+
+                    for (auto& target : targets) {
+                        ClusterBlockId bound = place_ctx.grid_blocks[target.second.x][target.second.y].blocks[target.second.sub_tile];
+                        if (bound != EMPTY_BLOCK_ID) unbindTile(target.second);
+                        bindTile(target.second, target.first);
+                        swaps_made.emplace_back(target.second, bound);
+                    }
+
+                    if (false) {
+                    fail:
+                        for (auto& swap : swaps_made) {
+                            unbindTile(swap.first);
+                            if (swap.second != EMPTY_BLOCK_ID)
+                                bindTile(swap.first, swap.second);
+                        }
+                        continue;
+                    }
+                    for (auto& target : targets) {
+                        ap->blk_locs[target.first].loc = target.second;
+                    }
+                    for (auto& swap : swaps_made) {
+                        if (swap.second != EMPTY_BLOCK_ID)
+                            remaining.emplace(imacro(swap.second) != -1 ? place_ctx.pl_macros[imacro(swap.second)].members.size() : 1, swap.second);
+                    }
+
+                    placed = true;
+                    break;
+                }
+            }
+        }
+    }
```

Comment:
This routine is about 240 lines long, without inline comments on what it's doing.
1. It should be factored into multiple smaller routines so the overall structure is more clear and you get out of these long loop nests, and avoid things like goto's.
2. It need inline comments (will be helped by calling sub-methods or sub-routines with meaningful names).